### ISSUE TITLE
proxy.autoconfig and security.OCSP.require (1212)

### ISSUE THREAD

**Issue by dvalter** [ghacksuserjs/ghacks-user.js#619]
#### Description

PAC scripts does not work (Firefox does not use it at all) when `security.OCSP.require` (1212) is set to `true`.

After I'd got it with the real PAC script, I used a pretty simple one to test with:
```js
function FindProxyForURL(url, host) {
    return "SOCKS5 127.0.0.1:9050";
}
```

#### Steps to reproduce:
1. Set security.OCSP.require to true
1. Add a valid URL containing code above to Preferences->Network Settings->Automatic proxy configuration URL (I've used valid public domain with HTTPS)
1. Start some Socks5 on localhost:9050 (e.g. tor daemon)
1. Try to load some show-my-ip page, you should see your real IP
1. Set security.OCSP.require to false
1. Click reload at Preferences->Network Settings->Automatic proxy configuration URL
1. Reload show-my-ip page, you should see your proxy exit IP (e.g. Tor exit node address)

#### Tested with:
GNU/Linux, amd64:
* Firefox 64, Arch and Fedora builds
* IceCat 60 (based on FF 60 ESR)
* Firefox 57
* Waterfox 56 (FF 56 fork)
_including default profiles with empty user.js_

According to #295 OCSP preference to hard-fail may cause some more proxy-related issues, so it may save plenty of time, if related to security.OCSP.require comments will contain some kind of related `[WARNING]` or `[BREAKAGE]` tag ( #545).

_Also I'm not sure it this issue is worth reporting to Mozilla._

**Comment by Thorin-Oakenpants** [comment]
Thanks. I'll address this later .. just got thru writing an entire book in a new topic and I'm feeling exhausted and I need some coffee

**Comment by Thorin-Oakenpants** [comment]
> > There's a reason FF has a default true.
> 
> But no, the FF default is `false`. At least on FF 64.0.2 .

Sorry, fucking overworked .. I did say I was exhausted and needed coffee. Didn't help that my value is user set since it's set from the user.js's overrides.

The default is false = no breakage = there's a reason why FF has it at that.

> Do you might have any public example that break when true ?

No, because I haven't had the pref at true for ages. And except for a few (major) sites like here at GitHub where I login and post shit .. all other sites are reading only (and default JS disabled), so it's not like I'm too worried about it all

**Comment by earthlng** [comment]
Thanks for the report.

> (I've used valid public domain with HTTPS)

are you sure the problem doesn't lie there? I mean afaik PAC scripts are usually loaded from a network share in enterprise environments. Using a public domain for a proxy script seems kind of weird. Maybe the OCSP server for that domain didn't respond?

It's probably worth opening a bugzilla ticket but I fear that they'll just remove the OCSP.require pref if there is indeed some kind of problem there. So yeah, IDK.

**Comment by dvalter** [comment]
> I actually have this in my overrides to set as false, I found it broke too many websites. There's a reason FF has a default true false (sorry, overworked as usual). It's on the list in #605 which will be used to help decide on any prefs to change to inactive, and also to finish off `[SETUP-`* tags.
> 
> This could definitely do with a `[SETUP-WEB]` tag .. maybe something like
> 
> ```js
> /* 1212: set OCSP fetch failures (non-stapled, see 1211) to hard-fail [SETUP-WEB]
>  * blah blah
>  * [SETUP-WEB] When true, affects PAC scripts and may cause other proxy-related issues
>  * [1] blah blah ***/
> user_pref("security.OCSP.require", true);
> ```
> 
> @dvalter does that seem like enough (I don't think we need to state the obvious that sites can hard-fail)?

That's absolutely fine.
Thank you for your reaction on this issue.

**Comment by Atavic** [comment]
It's the server providing the PAC File that's not OCSP aware. The same PAC File loaded locally, from localhost, will work fine.

**Comment by dvalter** [comment]
> It's the server providing the PAC File that's not OCSP aware. The same PAC File loaded locally, from localhost, will work fine.

Thank you for explaining. Both localhost HTTP and remote HTTPS with enabled OCSP stapling works fine.
But I still can't understand why exactly it's designed like this as far as CA (Let's Encrypt in my case) is accessible and regular server access does no cause any non-stapled OCSP hard-fail.

**Comment by dvalter** [comment]
> Thanks for the report.
> 
> > (I've used valid public domain with HTTPS)
> 
> are you sure the problem doesn't lie there? I mean afaik PAC scripts are usually loaded from a network share in enterprise environments. Using a public domain for a proxy script seems kind of weird. Maybe the OCSP server for that domain didn't respond?
> 
> It's probably worth opening a bugzilla ticket but I fear that they'll just remove the OCSP.require pref if there is indeed some kind of problem there. So yeah, IDK.

There's another rare usecase. It's bypassing some network restrictions and limitations without breaking local services and complex routing tricks. For example it is used against Internet censorship in Russia.

**Comment by Thorin-Oakenpants** [comment]
@earthlng bump .. what do we add, if any (it is in our overrides issue to look at) for a `[SETUP-*` tag - given that the problem from OP was it was loaded from a web request

**Comment by earthlng** [comment]
@dvalter 

> There's another rare usecase. It's bypassing some network restrictions and limitations without breaking local services and complex routing tricks. For example it is used against Internet censorship in Russia.

... and you need a PAC script for that? Why not configure the proxy directly?

@Thorin-Oakenpants 

IDK, it seems to be working as expected. I mean it doesn't make a lot of sense to setup a PAC script on a remote server because normally (in most network environments) you need that proxy script for every connection including that very 1st one to the public server where the PAC script is hosted. If that request is blocked somehow or maybe times out because OCSP is set to hard fail then yeah, you won't be able to use that script to setup your proxy settings. But if you *can* ie are allowed to change the PAC url then you might as well just specify the proxy directly. I don't really understand why he needs/wants to use a PAC script in the 1st place. If for some reason it *has* to be a PAC then just use a network share or maybe file URI.
On the other hand there's probably a bug here in that IMO it shouldn't fallback to a direct connection if a PAC script url is specified but FF fails to retrieve it.

**Comment by Atavic** [comment]
The PAC script contains encrypted urls that are read by the vpn to bypass gov. bans.

**Comment by Thorin-Oakenpants** [comment]
https://www.mozilla.org/en-US/security/advisories/mfsa2019-01/

There's one security fix in that lot for PACs if you're interested. Not saying it's related to this, but thought you might like to know

**Comment by Atavic** [comment]
@earthlng See: https://github.com/intika/Librefox/issues/106#issuecomment-457828517

**Comment by Atavic** [comment]
@Thorin-Oakenpants Bug is private, but to me it seems like it's related to localhost being resolved to both 127.0.0.1 and ::1 if the query doesn't specify one or the other. I use 127.0.0.1 as I avoid any ipv6 connection.

**Comment by Thorin-Oakenpants** [comment]
closing this with no changes (can revisit it in #605 for adding a breakage tag if needed) - the hard fail is just doing it's job: AFA-we-CT, it could be a FF bug if anyone wants to poke the beer :beer:  .. opps i meant poke the BEAR :bear: .... also (IANAExpert) it is unusual to load a PAC over like that (over https?)

Thanks @dvalter for the edge case :kiss:
